languages/julia/exercises/concept/poetry-club/.meta/design.md

```diff
@@ -0,0 +1,68 @@
+## Learning objectives
+
+- Basic string functions like `titlecase`, `replace` etc.
+- Be aware that unicode strings are the default and that the above mentioned functions will work on non-ascii input.
+- Know that strings can be treated as a collection of characters, e.g. for indexing.
+- How to concatenate strings.
+  - That and why `*` (and not `+`) is used for string concat.
+
+## Out of scope
+
+- Graphemes.
```

Comment:
After thinking about it some more, maybe the exercise should cover these to avoid teaching bad habits of indexing or iterating strings. Not sure.

Reply:
I think a real implementation of this function would have to use graphemes for correctness. This is what the implementation would look like:

```julia
using Unicode: graphemes

front_response(line) = titlecase(first(graphemes(line)))

back_response(line) = last(graphemes(remove_punctuation(line)))
```

And each of those would return a string.

Reasons why we might want to do this:

1. A line might end with a character and a combining character and the non-grapheme approach would select the wrong character. E.g. "Θεού" below is normalised, so each character is one unicode codepoint, but the last character could be equivalently expressed as ["\u3c5\u301"](https://en.wiktionary.org/wiki/%CF%8D), in which case `s[end]` would select the acute accent combining mark, not the upsilon.
2. Emojis
3. Arabic of Devanagari text, etc.

Reasons not to do this:

1. Sometimes we legitimately want to operate on characters (codepoints) or codeunits (bytes). In fact, most of the other string processing in this course is more correctly done on codepoints, I think.
2. It undermines the learning objective that strings are collections of characters (though it gives a different lesson, obv).
3. Using lots of short strings is kind of a pain.
4. The built-in Julia functions like `replace` are not grapheme-aware (though you will rarely notice this).

I don't know. A correct implementation using graphemes isn't too hard to understand, but I think we would have to hand-hold the student a bit and explain the difference between graphemes and characters and show them the exact functions to use.

It also exposes students to iterators, which are probably a bit odd to understand when they probably don't even know how to `collect` them yet.

Reply:
> I don't know. A correct implementation using graphemes isn't too hard to understand, but I think we would have to hand-hold the student a bit and explain the difference between graphemes and characters and show them the exact functions to use.

What do you think about removing the unicode aspect from this exercise entirely and then creating a 2nd part that extends it to work with unicode/graphemes? Story-wise it would be easy to add something like "Some famous poet from Greece/Russia is visiting..."

We'd basically have two concepts then:
- Strings as collection of characters/codepoints.
- What humans perceive as one character can consist of more than one character in code.

I don't like that it implicitely gives two options when in practice you're probably better off to treat every string as if it can contain graphemes unless you know with certainty that it's disallowed by the format of whatever you're reading.

> It also exposes students to iterators, which are probably a bit odd to understand when they probably don't even know how to collect them yet.

They might have seen them before, since loops will likely be a pre-req. While you don't strictly need a loop to solve this exercise, it's likely what many people would use.

Reply:
Testing this, `last(graphemes(s))` doesn't work. And neither does `first(Iterators.reverse(graphemes(s)))`. Turns out that's because random access with the UAX29 graphemes algorithm is [complicated](https://unicode.org/reports/tr29/).

`last(collect(graphemes(s)))` is probably the best you can do without a fair amount of work.

Reply:
I think it's realistic to present strings as collections of characters rather than collections of graphemes: that is how they are treated in most programs and it is only rarely problematic.

So I would be in favour of splitting graphemes out into a separate exercise. We should still present unicode input to students here, just not stuff that needs to be handled with `graphemes`.

Reply:
> I think a real implementation of this function would have to use graphemes for correctness.

> So I would be in favour of splitting graphemes out into a separate exercise. We should still present unicode input to students here, just not stuff that needs to be handled with `graphemes`.

After thinking a bit about the issue in [my other comment](https://github.com/exercism/v3/pull/2944#discussion_r547893463), I think that maybe we could use this exercise for graphemes, and then write another exercise about characters and strings being collections of them.

That allows proper implementations for this exercise that won't break with graphemes which as you say is closer to a real implementation.

I have a story idea in mind about an exercise where one has to determine if a set of characters meets certain requirements by chaining some `is...(c::Char)` methods. The final task of that could introduce strings as collections of characters. I'll have to write it out to know if it works though.

Reply:
Sounds good to me!

Reply:
Just jumping in here from where I think this exercise lives: I _purposefully_ didn't introduce codepoints outside of UTF-7/ASCII, _because_ it was my intention to have that in a different exercise.